File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/economy/BalanceFilter.java
package top.likoslupus.cellulosesz.api.economy;

import java.math.BigDecimal;
import java.util.Optional;

import static java.util.Objects.requireNonNull;

public record BalanceFilter(
        Optional<BigDecimal> minimum,
        Optional<BigDecimal> maximum
) {

    public BalanceFilter {
        requireNonNull(minimum, "minimum");
        requireNonNull(maximum, "maximum");
        if (minimum.isPresent() && maximum.isPresent()
                && minimum.orElseThrow().compareTo(maximum.orElseThrow()
        ) > 0) {
            throw new IllegalArgumentException("minimum must not exceed maximum");
        }
    }

    public static BalanceFilter all() {
        return new BalanceFilter(Optional.empty(), Optional.empty());
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/economy/BalanceFilter.kt
package top.likoslupus.cellulosesz.api.economy

import java.math.BigDecimal

@JvmRecord
public data class BalanceFilter(
    public val minimum: BigDecimal?,
    public val maximum: BigDecimal?
) {

    init {
        require(!(minimum != null && maximum != null && minimum > maximum)) {
            "minimum must not exceed maximum"
        }
    }

    public companion object {

        @JvmStatic
        public fun all(): BalanceFilter = BalanceFilter(null, null)

    }

}
